pkg/device/sync.go

```diff
@@ -127,7 +160,13 @@ func Sync(ctx context.Context, nodeID directpvtypes.NodeID) error {
 				if syncDrive(drive, devices[0]) {
 					updated = true
 				}
-
+				driveStateUpdated, err := syncVolume(ctx, drive)
```

Comment:
We need to have a separate function outside of drive sync up something like below

```go
func updateDrive(ctx context.Context, driveName string, deviceMap map[string][]device) error {
	drive, err := client.DriveClient().Get(ctx, driveName, metav1.GetOptions{})
	if err != nil {
		if apierrors.IsNotFound(err) {
			return nil
		}
		return err
	}
	var updated bool
	devices := deviceMap[drive.Status.FSUUID]
	switch len(devices) {
	case 0:
		// no match
		if drive.Status.Status != directpvtypes.DriveStatusLost {
			updated = true
			drive.Status.Status = directpvtypes.DriveStatusLost
		}
	case 1:
		// device found
		if syncDrive(drive, devices[0]) {
			updated = true
		}

		// verify mount
		switch drive.Status.Status {
		case directpvtypes.DriveStatusReady, directpvtypes.DriveStatusLost, directpvtypes.DriveStatusError, directpvtypes.DriveStatusMoving:
			source := utils.AddDevPrefix(string(drive.GetDriveName()))
			target := types.GetDriveMountDir(drive.Status.FSUUID)
			if err = xfs.Mount(source, target); err != nil {
				updated = true
				drive.Status.Status = directpvtypes.DriveStatusError
				drive.SetMountErrorCondition(fmt.Sprintf("unable to mount; %v", err))
				// Events and logs
				client.Eventf(drive, client.EventTypeWarning, client.EventReasonDriveMountError, "unable to mount the drive; %v", err)
				klog.ErrorS(err, "unable to mount the drive", "Source", source, "Target", target)
			} else {
				client.Eventf(
					drive,
					client.EventTypeNormal,
					client.EventReasonDriveMounted,
					"Drive mounted successfully to %s", target,
				)

				if drive.Status.Status == directpvtypes.DriveStatusLost {
					updated = true
					drive.Status.Status = directpvtypes.DriveStatusReady
				}

				latestErrorConditionType := drive.GetLatestErrorConditionType()
				if drive.Status.Status == directpvtypes.DriveStatusError {
					switch latestErrorConditionType {
					case directpvtypes.DriveConditionTypeMountError, directpvtypes.DriveConditionTypeMultipleMatches, directpvtypes.DriveConditionTypeRelabelError:
						updated = true
						drive.Status.Status = directpvtypes.DriveStatusReady
					}
				}

				if drive.Spec.Relabel {
					updated = true

					if err = os.Symlink(".", types.GetVolumeRootDir(drive.Status.FSUUID)); err != nil {
						switch {
						case errors.Is(err, os.ErrExist):
							err = nil
						default:
							if latestErrorConditionType != directpvtypes.DriveConditionTypeIOError {
								drive.Status.Status = directpvtypes.DriveStatusError
								drive.SetRelabelErrorCondition(fmt.Sprintf("unable to relabel; %v", err))
							}

							client.Eventf(
								drive,
								client.EventTypeWarning,
								client.EventReasonDriveRelabelError,
								"unable to relabel; %v", err,
							)

							klog.ErrorS(
								err,
								"unable to create symlink",
								"symlink", types.GetVolumeRootDir(drive.Status.FSUUID),
								"drive", drive.Name,
							)
						}
					}

					if err == nil {
						drive.Spec.Relabel = false
					}
				}
			}
		}
	default:
		// more than one matching devices
		updated = true
		var deviceNames string
		for i := range devices {
			if deviceNames != "" {
				deviceNames += ", "
			}
			deviceNames += devices[i].Name
		}
		drive.Status.Status = directpvtypes.DriveStatusError
		drive.SetMultipleMatchesErrorCondition(fmt.Sprintf("multiple devices found by FSUUID; %v", deviceNames))
		client.Eventf(drive, client.EventTypeWarning, client.EventReasonDriveHasMultipleMatches, "unable to mount the drive due to %v", err)
		klog.ErrorS(err, "multiple devices found by FSUUID", "drive", drive.GetDriveName(), "FSUUID", drive.Status.FSUUID, "devices", deviceNames)
	}
	if updated {
		_, err = client.DriveClient().Update(ctx, drive, metav1.UpdateOptions{TypeMeta: types.NewDriveTypeMeta()})
	}
	return err
}

// syncDrives - matches and syncs drives with locally probed devices.
func syncDrives(ctx context.Context, nodeID directpvtypes.NodeID) error {
	deviceMap, err := probeDeviceMap()
	if err != nil {
		return err
	}
	drives, err := drive.NewLister().NodeSelector([]directpvtypes.LabelValue{directpvtypes.ToLabelValue(string(nodeID))}).Get(ctx)
	if err != nil {
		return err
	}
	for i := range drives {
		if err = retry.RetryOnConflict(retry.DefaultRetry, func() error { return updateDrive(ctx, drives[i].Name, deviceMap) }); err != nil {
			return err
		}
	}

	return nil
}

// syncVolumes - matches and syncs drives with locally probed devices.
func syncVolumes(ctx context.Context, nodeID directpvtypes.NodeID) error {
	// FIXME
	return nil
}

// Sync - syncs drives and volumes
func Sync(ctx context.Context, nodeID directpvtypes.NodeID) error {
	derr := syncDrives(ctx, nodeID)
	verr := syncVolumes(ctx, nodeID)
	if derr != nil && verr != nil {
		return fmt.Errorf("%w; %w", derr, verr)
	}
	if derr != nil {
		return derr
	}
	return verr
}
```

Also move `sync.go` outside of `pkg/device` may be `pkg/sync`.

Reply:
moved this to volume listener Add()